mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -353,248 +427,293 @@ impl RingBuffer {
 
 impl Drop for RingBuffer {
     fn drop(&mut self) {
-        let result = self.mmap.flush();
+        let result = self.file.flush();
         if let Some(err) = result.err() {
             error!("Failed to flush {:?}", err);
         }
+        let result = self.metadata_file.flush();
+        if let Some(err) = result.err() {
+            error!("Failed to flush metadata {:?}", err);
+        }
     }
 }
 
-fn create_file(file_path: &Path) -> IOResult<File> {
+fn create_file(file_path: &PathBuf) -> IOResult<File> {
+    if let Some(parent) = file_path.parent() {
+        create_dir_all(parent)?;
+    }
     OpenOptions::new()
         .read(true)
         .write(true)
         .create(true)
         .open(file_path)
 }
 
+fn save_ring_buffer_metadata(metadata: &RingBufferMetadata, file: &mut File) -> BincodeResult<()> {
+    let buf = bincode::serialize(metadata)?;
+    file.seek(SeekFrom::Start(0))?;
+    file.write_all(&buf)?;
+    file.flush()?;
+    Ok(())
+}
+
+fn retrieve_ring_buffer_metadata(file: &mut File) -> BincodeResult<RingBufferMetadata> {
+    let mut buf = vec![];
+    file.read_to_end(&mut buf)?;
+    if buf.is_empty() {
+        return Ok(RingBufferMetadata::default());
+    }
+    bincode::deserialize::<RingBufferMetadata>(&buf)
+}
+
 fn find_pointers_and_order_post_crash(
-    mmap: &MmapMut,
-    max_file_size: u32,
-) -> StorageResult<(FilePointers, u128)> {
-    let mut block: BlockHeaderWithHash;
+    // mmap: &MmapMut,
+    file: &mut File,
+    max_file_size: u64,
+    best_guess_metadata: RingBufferMetadata,
+) -> RingBufferMetadata {
+    let block_size = *SERIALIZED_BLOCK_SIZE;
+
+    let mut start = best_guess_metadata.file_pointers.write;
```

Comment:
We should only need to do half the work as what this function did previously. This is because the metadata, if any, found us a location of a block close to where we left off.